apps/omg/lib/omg/output.ex

```diff
@@ -14,53 +14,108 @@
 
 defmodule OMG.Output do
   @moduledoc """
-  `OMG.Output` and `OMG.Output.Protocol` represent the outputs of transactions, i.e. the valuables or other pieces of
+  `OMG.Output` represent the outputs of transactions, i.e. the valuables or other pieces of
   data spendable via transactions on the child chain, and/or exitable to the root chain.
 
   This module specificially dispatches generic calls to the various specific types
   """
 
-  @output_types_modules Application.fetch_env!(:omg, :output_types_modules)
-  @type_markers Map.keys(@output_types_modules)
+  @type t :: %__MODULE__{
+          owner: OMG.Crypto.address_t(),
+          currency: OMG.Crypto.address_t(),
+          amount: non_neg_integer()
+        }
 
-  def dispatching_reconstruct([type_marker | raw_rlp_decoded_chunks]) when type_marker in @type_markers do
-    protocol_module = @output_types_modules[type_marker]
-    protocol_module.reconstruct(raw_rlp_decoded_chunks)
-  end
+  @output_type_marker <<1>>
 
-  def dispatching_reconstruct(_), do: {:error, :unrecognized_output_type}
+  defstruct [:owner, :currency, :amount]
 
-  def from_db_value(%{type: type_marker} = db_value), do: @output_types_modules[type_marker].from_db_value(db_value)
-  # default clause for backwards compatibility
-  def from_db_value(%{} = db_value), do: OMG.Output.FungibleMoreVPToken.from_db_value(db_value)
-end
+  # TODO(achiurizo)
+  # Need to fix that this method is able to re-build from it's own generated rlp data.
+  # ex: [<<1>>, <<1::160>>, <<1::160>>, 1] (last number is an integer instead of binary)
+  @doc """
+  Returns a OMG.Output struct from a map
 
-defprotocol OMG.Output.Protocol do
```

Comment:
This is behaviour I want to move into ex_plasma. Secondly I disagree that protocols is the right approach here(why not just use a [behaviour?](https://github.com/omisego/ex_plasma/blob/master/lib/ex_plasma/transaction.ex#L12) You can extend [additional types like payment transactions](https://github.com/omisego/ex_plasma/blob/master/lib/ex_plasma/transactions/payment.ex#L11). The dispatching happens [all within the module](https://github.com/omisego/ex_plasma/blob/master/lib/ex_plasma/transaction.ex#L129).

While the current ex_plasma has implemented such dispatching for output_type, to implment it with the current setup won't be difficult.

Reply:
"future code", "readyiness for future code", always a bad sign. don't ever do it if you can avoid it.

Reply:
> So, please, let's try to keep the critique constructive and considerate.

If it comes off that our discussion are not constructive or considerate, then I apologize. Discussions like these can be contentious so that we can understand the motivations behind it and how to move forward.

>  It is hard to see the plan, if it lives in a separate repo.)

I agree, but I that's the separation of concerns I think we need to move towards. The responsibility of how to interact with the contract is making it much more difficult for us to focus on the other responsibilities of the child chain and watcher.

> BTW, why wouldn't we rather fix in elixir-omg and split ex_plasma out later or split out "as-is" and fix after split. It would be much easier to see the plan through.

This I accept responsibility. I wanted to move quickly to understand the inner workings of the encoding/decoding process and build a fuller test suite that I can run quickly with. I also didn't want to make such large changes that would affect the push towards ALD directly. By coming in with full unit tests et al, my goal was to be able to introduce these changes effectively without much disruption (beyond the removal of the code-paths that are doing in repo).